### PR TITLE
trigger approval actions from FE

### PR DESCRIPTION
**Pull request by bergarces**
## Explanation

- Updates assets-controllers `TokensController`, which creates the need to trigger the acceptance or rejection of the approval from the FE.
- A follow up PR will remove the interactions with the `TokensController` from the FE and just accept/reject the approval.
- Added E2E tests for the `wallet_watchAsset` action, as only manually adding the token was being tested.

Fixes https://github.com/MetaMask/MetaMask-planning/issues/453

## Manual Testing Steps

- Open Test Dapp
- Access EIP 747 section
- Try to add a new ERC20 token

## Pre-merge author checklist

- [X] I've clearly explained:
  - [X] What problem this PR is solving
  - [X] How this problem was solved
  - [X] How reviewers can test my changes
- [X] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

File: ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.js
Comment (bergarces) [inline ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.js:127]: I'm not all that sure about this `map` with an `async` function inside that awaitsthe `dispatch` but doesn't await `trackEvent`.

I considered returning the three promises with a `flatMap` instead, but I don't want to change existing behaviour so I'm going to assume there's a reason for it (such as not wanting to throw if `trackEvent` fails).

File: ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.js
Comment (bergarces) [inline ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.js:154]: Copied this error from the permissions confirmation screen.

https://github.com/MetaMask/metamask-extension/blob/develop/ui/pages/permissions-connect/permissions-connect.component.js#L383

File: ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.js
Comment (bergarces) [inline ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.js:129]: This will not execute if I just add another line with `await dispatch(acceptWatchAsset(id));`, I have to return the whole thing with a flatMap.

https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/45107/workflows/70f1b4d8-b672-4e04-984c-f22e56e3acad/jobs/1275053

Comment (OGPoyraz) [reply ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.js:129]: Nice one!

**Review by OGPoyraz — DISMISSED** [review]

**Comment by OGPoyraz** [comment]
I will approve again once patches removed 👍

File: test/e2e/tests/add-hide-token.spec.js
Comment (matthewwalsh0) [inline test/e2e/tests/add-hide-token.spec.js:237]: Should we be asserting something at this point? Such as that there is one less window, or that our original request promise was rejected?

File: ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.js
Comment (matthewwalsh0) [reply ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.js:129]: I know this code won't last long but would another option have been to dispatch the `acceptWatchAsset` with a `then` to chain it to the `resolvePendingApproval`?

It's probably redundant as there was no error handling here anyway but it would ensure we don't resolve the approval and hide the confirmation until the accept side effect is completed.

File: test/e2e/tests/add-hide-token.spec.js
Comment (bergarces) [reply test/e2e/tests/add-hide-token.spec.js:237]: I couldn't see what we could be asserting here. By pressing reject it just won't add the currency, but there's no e2e helper method that I could find to check that an element has not been added. The parent element is also not particularly helpful in this case (no clear list of elements underneath to check assert the length).

When I checked other e2e tests such as the add custom network, it just presses the cancel button and doesn't assert anything afterwards.

File: ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.js
Comment (bergarces) [reply ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.js:126]: I've been seeing how to sort it, and I can't see an easy way out.

Since `acceptWatchAsset` doesn't throw when it fails to add the asset and it doesn't return anything, I cannot wrap it on a try/catch statement to reject the pending approval instead.

File: ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.js
Comment (bergarces) [reply ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.js]: It turns out that the issue was that with the event handler being async, it was trying to assert things before they were being called. It's been sorted using `act` (https://legacy.reactjs.org/docs/testing-recipes.html#act).

File: test/e2e/tests/add-hide-token.spec.js
Comment (bergarces) [reply test/e2e/tests/add-hide-token.spec.js:237]: Added additional validation.

**Pull request closed without merging**
